File: go/providers/mailtm.go
package providers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"
)

const mailTMBaseURL = "https://api.mail.tm"

// MailTM uses mail.tm — bearer token auth via account creation.
type MailTM struct {
	client *HttpClient
	token  string
	config map[string]string
}

func NewMailTM(config map[string]string) *MailTM {
	proxies := ParseProxies(config["proxies"])
	randomUA := config["randomUa"] == "true"

	return &MailTM{
		client: NewHttpClient(proxies, randomUA, false),
		config: config,
		token:  config["token"],
	}
}

func (m *MailTM) GenerateEmail() (string, error) {
	password := m.config["password"]
	if password == "" {
		password = "TempPass123!"
	}

	// Get an available domain.
	domBody, err := m.client.Get(mailTMBaseURL + "/domains")
	if err != nil {
		return "", &TempMailError{Provider: "mailtm", Message: err.Error()}
	}

	var domains struct {
		HydraMember []struct {
			ID     string `json:"@id"`
			Domain string `json:"domain"`
		} `json:"hydra:member"`
	}
	if err := json.Unmarshal(domBody, &domains); err != nil || len(domains.HydraMember) == 0 {
		return "", &TempMailError{Provider: "mailtm", Message: "no domains available"}
	}
	domain := domains.HydraMember[0].Domain

	// Create random address.
	addr := fmt.Sprintf("tmp_%d@%s", time.Now().UnixNano()%1e9, domain)
	payload := map[string]string{
		"address":  addr,
		"password": password,
	}
	b, _ := json.Marshal(payload)

	_, err = m.client.Post(mailTMBaseURL+"/accounts", "application/json", bytes.NewReader(b))
	if err != nil {
		return "", &TempMailError{Provider: "mailtm", Message: err.Error()}
	}

	// Obtain token.
	loginPayload := map[string]string{
		"address":  addr,
		"password": password,
	}
	lb, _ := json.Marshal(loginPayload)
	tokenBody, err := m.client.Post(mailTMBaseURL+"/token", "application/json", bytes.NewReader(lb))
	if err == nil {
		var tokenData struct {
			Token string `json:"token"`
		}
		json.Unmarshal(tokenBody, &tokenData)
		if tokenData.Token != "" {
			m.token = tokenData.Token
		}
	}

	return addr, nil
}

func (m *MailTM) GetInbox(email string) ([]Message, error) {
	req, _ := http.NewRequest(http.MethodGet, mailTMBaseURL+"/messages", nil)
	req.Header.Set("Authorization", "Bearer "+m.token)
	req.Header.Set("Accept", "application/ld+json")

	body, err := m.client.RawDo(req, nil)
	if err != nil {
		return nil, &TempMailError{Provider: "mailtm", Message: err.Error()}
	}

	var raw struct {
		HydraMember []struct {
			ID   string `json:"@id"`
			From struct {
				Address string `json:"address"`
			} `json:"from"`
			Subject        string `json:"subject"`
			HasAttachments bool   `json:"hasAttachments"`
			CreatedAt      string `json:"createdAt"`
		} `json:"hydra:member"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, &TempMailError{Provider: "mailtm", Message: "parse error: " + err.Error()}
	}

	msgs := make([]Message, len(raw.HydraMember))
	for i, r := range raw.HydraMember {
		t, _ := time.Parse(time.RFC3339, r.CreatedAt)
		idParts := strings.Split(r.ID, "/")
		id := idParts[len(idParts)-1]
		msgs[i] = Message{
			ID:      id,
			Sender:  r.From.Address,
			Subject: r.Subject,
			Date:    t,
		}
	}
	return msgs, nil
}

func (m *MailTM) ReadMessage(messageID string) (*MessageDetail, error) {
	req, _ := http.NewRequest(http.MethodGet, mailTMBaseURL+"/messages/"+messageID, nil)
	req.Header.Set("Authorization", "Bearer "+m.token)
	req.Header.Set("Accept", "application/ld+json")

	body, err := m.client.RawDo(req, nil)
	if err != nil {
		return nil, &TempMailError{Provider: "mailtm", Message: err.Error()}
	}

	var raw struct {
		ID   string `json:"@id"`
		From struct {
			Address string `json:"address"`
		} `json:"from"`
		Subject     string   `json:"subject"`
		Text        string   `json:"text"`
		HTML        []string `json:"html"`
		CreatedAt   string   `json:"createdAt"`
		Attachments []struct {
			Filename string `json:"filename"`
			MIMEType string `json:"mimeType"`
			Size     int    `json:"size"`
		} `json:"attachments"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, &TempMailError{Provider: "mailtm", Message: "parse error: " + err.Error()}
	}

	t, _ := time.Parse(time.RFC3339, raw.CreatedAt)
	idParts := strings.Split(raw.ID, "/")
	id := idParts[len(idParts)-1]
	htmlBody := strings.Join(raw.HTML, "\n")

	detail := &MessageDetail{
		Message: Message{
			ID:      id,
			Sender:  raw.From.Address,
			Subject: raw.Subject,
			Date:    t,
		},
		BodyText: raw.Text,
		BodyHTML: htmlBody,
	}
	for _, a := range raw.Attachments {
		detail.Attachments = append(detail.Attachments, map[string]any{
			"filename":     a.Filename,
			"content_type": a.MIMEType,
			"size":         a.Size,
		})
	}
	return detail, nil
}

func (m *MailTM) DeleteEmail(email string) (bool, error) {
	return false, &TempMailError{Provider: "mailtm", Message: "mailbox deletion not supported, delete individual messages"}
}

func (m *MailTM) WaitForEmail(email string, timeout time.Duration, interval time.Duration) (*Message, error) {
	return waitForEmail(m, email, timeout, interval)
}

File: javascript/src/providers/MailTm.js
import { TempMailProvider } from '../base.js';
import { Message, MessageDetail } from '../models.js';
import { TempMailError, NotFoundError } from '../errors.js';

const API = 'https://api.mail.tm';

/**
 * mail.tm provider — bearer token auth.
 */
export class MailTm extends TempMailProvider {
  /** @param {{ password?: string, randomUA?: boolean, useCookies?: boolean }|undefined} config */
  constructor(config = {}) {
    super('mail.tm', { randomUA: config.randomUA, useCookies: config.useCookies });
    this.password = config.password || 'P@ssw0rd!Temp';
    /** @type {string|null} */
    this._token = null;
    /** @type {string|null} */
    this._accountId = null;
  }

  /** @returns {Promise<string>} */
  async generateEmail() {
    return await this._generateEmailCore();
  }

  /** @returns {Promise<string>} */
  async _generateEmailCore() {
    const domainsRes = await this._fetch(`${API}/domains`);
    const domains = await domainsRes.json();
    const domain = domains['hydra:member']?.[0]?.domain || domains.member?.[0]?.domain;
    if (!domain) throw new TempMailError('No domain available', this.name);

    const address = `tmp${Date.now().toString(36)}@${domain}`;
    const res = await this._fetch(`${API}/accounts`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ address, password: this.password }),
    });
    const account = await res.json();
    this._accountId = account.id;

    // Get token
    const tokenRes = await this._fetch(`${API}/token`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ address, password: this.password }),
    });
    const tokenData = await tokenRes.json();
    this._token = tokenData.token;

    return address;
  }

  /** @returns {Promise<Record<string, string>>} */
  _headers() {
    const h = { 'Content-Type': 'application/json' };
    if (this._token) h['Authorization'] = `Bearer ${this._token}`;
    return h;
  }

  /**
   * @param {string} _email
   * @returns {Promise<Message[]>}
   */
  async getInbox(_email) {
    const res = await this._fetch(`${API}/messages`, { headers: this._headers() });
    const data = await res.json();
    const items = data['hydra:member'] || data.member || [];
    return items.map(m => new Message({
      id: m.id, // raw ID — MailTm uses token auth, no email context needed
      sender: m.from?.address || m.from?.name || 'unknown',
      subject: m.subject || '',
      date: m.createdAt,
    }));
  }

  /**
   * @param {string} messageId
   * @returns {Promise<MessageDetail>}
   */
  async readMessage(messageId) {
    const res = await this._fetch(`${API}/messages/${messageId}`, { headers: this._headers() });
    const m = await res.json();
    if (!m.id) throw new NotFoundError(`Message ${messageId} not found`, this.name);
    return new MessageDetail({
      id: m.id,
      sender: m.from?.address || m.from?.name || 'unknown',
      subject: m.subject || '',
      date: m.createdAt,
      bodyText: m.text || '',
      bodyHtml: m.html?.[0] || '',
      attachments: (m.attachments || []).map(a => ({
        filename: a.filename,
        contentType: a.contentType,
        url: `${API}/sources/${m.id}/attachments/${a.id}`,
      })),
    });
  }

  /**
   * @param {string} _email
   * @returns {Promise<boolean>}
   */
  async deleteEmail(_email) {
    if (!this._accountId) return false;
    await this._fetch(`${API}/accounts/${this._accountId}`, {
      method: 'DELETE',
      headers: this._headers(),
    });
    this._token = null;
    this._accountId = null;
    return true;
  }
}
